crates/concrete_check/src/linearity_check.rs

```diff
@@ -0,0 +1,1086 @@
+use concrete_ast::Program;
+//use concrete_ast::modules::{Module, ModuleDefItem};
+use concrete_ast::functions::{FunctionDecl, Param};
+use concrete_ast::modules::ModuleDefItem;
+//use concrete_ast::functions::FunctionDef;
+use concrete_ast::expressions::{Expression, PathOp, StructInitField, ValueExpr};
+//use concrete_ast::statements::{Statement, AssignStmt, LetStmt, WhileStmt, ForStmt, LetStmtTarget, Binding};
+use concrete_ast::statements::{AssignStmt, Binding, LetStmt, LetStmtTarget, Statement};
+
+use concrete_ast::types::TypeSpec;
+//use concrete_ast::expressions::Value; // Import the missing module
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+enum VarState {
+    Unconsumed,
+    Consumed,
+    _Borrowed,
+    _BorrowedMut,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct VarInfo {
+    ty: String, //TODO Define 'Type' as a struct or enum?
+    depth: usize,
+    state: VarState,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+enum CountResult {
+    Zero,
+    One,
+    MoreThanOne,
+}
+
+#[derive(Debug, Clone, Copy)]
+struct Appearances {
+    consumed: u32,
+    write: u32,
+    read: u32,
+    path: u32,
+}
+
+// TODO remove. This are structures translated from Austral
+#[allow(dead_code)]
+enum Expr {
+    NilConstant,
+    BoolConstant(bool),
+    IntConstant(i32),
+    FloatConstant(f64),
+    StringConstant(String),
+    ConstVar,
+    ParamVar(String),
+    LocalVar(String),
+    FunVar,
+    Funcall(Box<Expr>, Vec<Expr>),
+    MethodCall(Box<Expr>, Vec<Expr>),
+    VarMethodCall(Vec<Expr>),
+    FptrCall(Box<Expr>, Vec<Expr>),
+    Cast(Box<Expr>, String),
+    Comparison(Box<Expr>, Box<Expr>),
+    Conjunction(Box<Expr>, Box<Expr>),
+    Disjunction(Box<Expr>, Box<Expr>),
+    Negation(Box<Expr>),
+    IfExpression(Box<Expr>, Box<Expr>, Box<Expr>),
+    RecordConstructor(Vec<Expr>),
+    UnionConstructor(Vec<Expr>),
+    Path { head: Box<Expr>, elems: Vec<Expr> },
+    Embed(Vec<Expr>),
+    Deref(Box<Expr>),
+    SizeOf,
+    Borrow(BorrowMode, String),
+    ArrayIndex(Box<Expr>),
+}
+
+#[allow(dead_code)]
+enum BorrowMode {
+    ReadBorrow,
+    WriteBorrow,
+}
+
+impl Appearances {
+    fn new(consumed: u32, write: u32, read: u32, path: u32) -> Self {
+        Appearances {
+            consumed,
+            write,
+            read,
+            path,
+        }
+    }
+
+    fn partition(count: u32) -> CountResult {
+        match count {
+            0 => CountResult::Zero,
+            1 => CountResult::One,
+            _ => CountResult::MoreThanOne,
+        }
+    }
+
+    fn zero() -> Self {
+        Self::new(0, 0, 0, 0)
+    }
+
+    fn consumed_once() -> Self {
+        Self::new(1, 0, 0, 0)
+    }
+
+    // When borrowed implemented
+    fn _read_once() -> Self {
+        Self::new(0, 0, 1, 0)
+    }
+
+    fn _write_once() -> Self {
+        Self::new(0, 1, 0, 0)
+    }
+
+    fn path_once() -> Self {
+        Self::new(0, 0, 0, 1)
+    }
+
+    /* TODO implementation of merge without copying
```

Comment:
can you summarize here in the TODO what is needed to complete the implementation or why it is commented?

Reply:
The approach was an intended optimization, but it is not necessary by now as we are far from optimization stage. Removed the commented code

```
    fn merge(&mut self, other: &Appearances) {
            self.consumed += other.consumed;
            self.write += other.write;
            self.read += other.read;
            self.path += other.path;
    }
```